File: java/algorithms/sort/QuickSort.java
package sort;

import java.util.Arrays;

import org.junit.Test;

public class QuickSort {

	public void arrSwap(int[] arr, int posA, int posB) {
		int tmp = arr[posA];
		arr[posA] = arr[posB];
		arr[posB] = tmp;
	}

	public int partition(int[] arr, int start, int end) {
		int pivotPos = start;
		int pivot = arr[start];
		for (int i = start; i <= end; i++) {
			if (arr[i] < pivot) {
				pivotPos++;
				arrSwap(arr, pivotPos, i);
			}
		}
		arrSwap(arr, start, pivotPos);
		return pivotPos;
	};

	public void quicksort(int[] arr, int start, int end) {
		if (start < end) {
			int pivotPos = partition(arr, start, end);
			quicksort(arr, start, pivotPos - 1);
			quicksort(arr, pivotPos + 1, end);
		}
	}

	@Test
	public void testQuicksort() {
		int[] arr = new int[] { 8, 2, 4, 65, 2, 4, 7, 1, 9, 0, 2, 34, 12 };
		quicksort(arr, 0, arr.length - 1);
		System.out.println(Arrays.toString(arr));
	}

	public void sort2(int[] arr, int l, int r) {
		if (l < r) {
			int i = l, j = r, x = arr[l];

			while (i < j) {
				while (i < j && arr[j] >= x) {
					j--;
				}
				if (i < j) {
					arr[i] = arr[j];
					i++;
				}
				while (i < j && arr[i] < x) {
					i++;
				}
				if (i < j) {
					arr[j] = arr[i];
					j--;
				}
			}

			arr[i] = x;
			sort2(arr, l, i - 1);
			sort2(arr, i + 1, r);
		}
	}
	
	@Test
	public void testSort2() {
		int[] arr = new int[] { 8, 2, 4, 65, 2, 4, 7, 1, 9, 0, 2, 34, 12 };
		sort2(arr, 0, arr.length - 1);
		System.out.println(Arrays.toString(arr));
	}
}

File: sort/quicksort.go
package sort

/*
	分割，快排左右
	分割：
	1 取轴
	2 中位为start（轴本身占1位），循环除轴的节点
	2.1 节点值小时，中位后移，交换节点和中位的值
	3 交换轴与中位的值
*/
func QuickSort(arr []int, start int, end int) {
	swap := func(i int, j int) {
		arr[i], arr[j] = arr[j], arr[i]
	}
	partition := func(start int, end int) int {
		pivot := start
		val := arr[start]

		pos := start
		for i := start + 1; i <= end; i++ {
			if arr[i] < val {
				pos++
				swap(pos, i)
			}
		}
		swap(pivot, pos)
		return pos
	}
	if start < end {
		pivot := partition(start, end)
		QuickSort(arr, start, pivot-1)
		QuickSort(arr, pivot+1, end)
	}
}

/*
	..
	双指针法分割
	1 取轴为开头（或交换到开头），开头为左点，缓存值
	2 向中循环
	2.1 找到右点，向左点赋值
	2.2 找到左点，向右点赋值
	3. 找到中点，赋缓存值
*/
func QuickSort2(arr []int, start int, end int) {
	partition := func(start int, end int) int {
		i := start
		j := end
		val := arr[start]

		for i < j {
			for i < j && arr[j] >= val {
				j--
			}
			if i < j {
				arr[i] = arr[j]
				i++
			}
			for i < j && arr[i] < val {
				i++
			}
			if i < j {
				arr[j] = arr[i]
				j--
			}
		}
		arr[i] = val
		return i
	}
	if start < end {
		pivot := partition(start, end)
		QuickSort2(arr, start, pivot-1)
		QuickSort2(arr, pivot+1, end)
	}
}
